master/buildbot/statistics/capture.py

```diff
@@ -0,0 +1,200 @@
+
+
+class Capture(object):
+    """
+    Base class for all Capture* classes.
+    """
+    def __init__(self, routingKey, callback):
+        self.routingKey = routingKey
+        self.callback = callback
+        # parent service and buildmaster to be set when StatsService initialized
+        self.parent_svcs = []
+        self.master = None
+
+    def defaultContext(self, msg):
+        return {
+            "builder_name": self.builder_name,
+            "build_number": str(msg['number'])
+        }
+
+    def consumer(self, routingKey, msg):
+        raise NotImplementedError
+
+
+class CaptureProperty(Capture):
+    """
+    Convenience wrapper for getting statistics for filtering.
+    Filters out build properties specifies in the config file.
+    """
+    def __init__(self, builder_name, property_name, callback=None):
+        self.builder_name = builder_name
+        self.property_name = property_name
+        routingKey = ("builders", None, "builds", None, "finished")
+
+        def default_callback(props, property_name):
+            return props[property_name][0]  # index: 0 - prop_value, 1 - prop_source
+
+        if not callback:
+            callback = default_callback
+
+        Capture.__init__(self, routingKey, callback)
+
+    @defer.inlineCallbacks
+    def consumer(self, routingKey, msg):
+        """
+        Consumer for this (CaptureProperty) class. Gets the properties from data api and
+        send them to the storage backends.
+        """
+        builder_info = yield self.master.data.get(("builders", msg['builderid']))
+        if self.builder_name == builder_info['name']:
+            properties = yield self.master.data.get(("builds", msg['buildid'], "properties"))
+            ret_val = self.callback(properties, self.property_name)
+            context = self.defaultContext(msg)
+            series_name = self.builder_name + "-" + self.property_name
+            post_data = {
+                "name": self.property_name,
+                "value": ret_val
+            }
+            for svc in self.parent_svcs:
+                yield threads.deferToThread(svc.postStatsValue, post_data, series_name,
+                                            context)
+
+        else:
+            yield defer.succeed(None)
+
+
+class CaptureBuildTimes(Capture):
+    """
+    Capture methods for capturing build start times.
+    """
+    def __init__(self, builder_name, callback):
+        self.builder_name = builder_name
+        routingKey = ("builders", None, "builds", None, "finished")
+        Capture.__init__(self, routingKey, callback)
+
+    @defer.inlineCallbacks
+    def consumer(self, routingKey, msg):
+        """
+        Consumer for CaptureBuildStartTime. Gets the build start time.
+        """
+        builder_info = yield self.master.data.get(("builders", msg['builderid']))
+        if self.builder_name == builder_info['name']:
+            ret_val = self.callback(*self.retValParams(msg))
```

Comment:
You expect `retValParams` method to be defined by children, please add it to this class and document expectations.